### PR TITLE
Add constrained planning tutorial

### PR DESCRIPTION
**Pull request by sjahr**
Run with
```
ros2 launch moveit_drake constrained_planning_demo.launch.py
```

### CODE REVIEW

File: demo/launch/constrained_planning_demo.launch.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+            ],
+        }
+    }
+
+    # Start the actual move_group node/action server
+    run_move_group_node = Node(
+        package="moveit_ros_move_group",
+        executable="move_group",
+        output="screen",
+        parameters=[moveit_config.to_dict(), drake_ktopt_planning_pipeline_config],
```
Comment (sjahr) [inline demo/launch/constrained_planning_demo.launch.py:55]: This doesn't work yet. @kamiradi I won't be able to look further into it, maybe you can fix it. Feel free to just push to this PR

File: demo/src/constrained_planning_demo.cpp [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+  plane_constraints.position_constraints.emplace_back(plane_constraint);
+  plane_constraints.name = "use_equality_constraints";
+  move_group_interface.setPathConstraints(plane_constraints);
+
+  // And again, configure and solve the planning problem
+  move_group_interface.setPoseTarget(target_pose);
+  move_group_interface.setPlanningTime(10.0);
+
+  // Use KTOpt
+  move_group_interface.setPlanningPipelineId("drake_ktopt");
```
Comment (sjahr) [inline demo/src/constrained_planning_demo.cpp:159]: Once KTOpt is loaded you should be able to configure it like this or with a function like this from the move_group interface API

**Comment by sjahr** [comment]
@kamiradi Good news: The planner is now loadable :tada: Bad news: KTOpt fails to plan in both demos --> Maybe the problem is overconstrained? I clean-ed up the position - jerk constraint parsing, feel free to double check but it should work but I fear we introduced and error somewhere along the line that we need to debug

**Comment by sea-bass** [comment]
Does it work if you remove the constraints?

**Comment by kamiradi** [comment]
It is most definitely the box constraints. I tried it with the jerk
constraints and the pipeline benchmark was working. I knew this was going
to be an issue, bit waited for the constrained planing Tut to work.

**Comment by sea-bass** [comment]
Sorry, I keep forgetting that my computer can't handle building MoveIt from source unless I add swap/zram and I just crashed mid build, then ran out of time to try again.

Things look largely okay, the one place I would check is whether using `plant.world_frame()` as the "frame A" vs. something like `"panda_link0"`, which is the base of the planning group? Or maybe those frames are coincident and it's something else.

**Comment by kamiradi** [comment]
Yeah no worries. I'll have a look tomorrow morning

**Comment by sea-bass** [comment]
I can only get this to work if:

1. I disable the Meschat visualization
2. I remove the position/velocity/acceleration/jerk constraints
3. I remove the path constraints

So I'm not really sure whether 2. or 3. is the problem here.

**Comment by kamiradi** [comment]
Having a look at this now

**Comment by sea-bass** [comment]
OK I found the issue with the joint bounds... using `-std::numeric_limits<double>::max()` for the unbounded joints (in this case, gripper) could be bad -- we should be using `std::numeric_limits<double>::lowest()`.

Also, the gripper joints were actually uninitialized to garbage values, which caused failures.

Furthermore, using these `std::numeric_limits` extremum values for the velocity, acceleration, and jerk also caused optimization to fail. I defined constant that set it to a large finite value (I'm using `100`) for all unbounded joints, and this makes things work.

So that clears issue 2 above... now for the constraints.

**Comment by sea-bass** [comment]
For issue 3, I was able to address this by padding the constraints to be more conservative at the optimizer level.

After fixing the above problem with joint limits, I realized that the optimization now succeeded with constraints, but failed on the validation step. This makes sense, as Drake only optimizes at the control points, and intermediate trajectory points could therefore violate constraints.

Adding a padding distance parameter to the `generate_parameter_library` YAML, I could get things working!

https://github.com/user-attachments/assets/30da6af8-b5d9-4192-9cc4-81df3ed25206

File: src/ktopt_planning_context.cpp [hunk 3]
```diff
@@ -215,29 +159,29 @@ void KTOptPlanningContext::solve(planning_interface::MotionPlanResponse& res)
...
+  //   plant.SetPositions(&plant_context, traj.value(t));
+  //   auto& vis_context = visualizer_->GetMyContextFromRoot(*diagram_context_);
+  //   visualizer_->ForcedPublish(vis_context);
+  //   // Without these sleeps, the visualizer won't give you time to load your
+  //   // browser
+  //   // TODO: This should not hold up planning time
+  //   std::this_thread::sleep_for(std::chrono::milliseconds(static_cast<int>(params_.trajectory_time_step * 10000.0)));
+  // }
+  // visualizer_->StopRecording();
+  // visualizer_->PublishRecording();
```
Comment (sea-bass) [inline src/ktopt_planning_context.cpp:177]: I commented this out BTW, see if this is necessary.

**Comment by sea-bass** [comment]
One last comment on this -- when I looked at Meshcat, the gripper fingers definitely fly all over the place (which might be a reason for failure, and maybe also why I had to bound the velocity/accel/jerk to finite values).

I think what we need to do is take all of the joints that are not actively part of this group, and constrain their positions to exactly the starting positions + velocities and further derivatives to zero. Will let someone else try that out.

**Comment by kamiradi** [comment]
Works for me as well :partying_face: . I added meshcat visualise to the parameter yaml as suggested. 

I'll track the panda finger as a separate issue and get to it. For now the bounding box constraint works, I'll add the orientation constraints and we can call it a ROSCon.

**Pull request merged**